### PR TITLE
Doctor plugin & Kubernetes plugin usage

### PR DESCRIPTION
**Pull request by huseyinbabal**
## Description

Changes proposed in this pull request:

- Added doctor plugin that can provide possible actions for errors
- Kubernetes & doctor plugin integration is added

## Testing

- Checkout and build plugin 
```bash
PLUGIN_DOWNLOAD_URL_BASE_PATH="http://localhost:8080/plugin-dist" make gen-plugins-index
```
- Serve plugins
```bash
npx serve --listen 8080
```
- Run botkube withfollowing config
```yaml
rbac:
  create: true
  rules:
    - apiGroups: ["*"]
      resources: ["*"]
      verbs: ["get", "watch", "list"]
  staticGroupName: &static-group-name "botkube-plugins-default"
communications:
  default-group:
    socketSlack:
      enabled: true
      channels:
        default:
          name: botkube-demo
          bindings:
            sources:
              - 'k8s-all-err-events'
              - 'k8s-recommendation-events'
              - 'k8s-all-events'
            executors:
              - ai
              - k8s-default-tools
      appToken: "xapp-1-"
      botToken: "xoxb-"

sources:
  'k8s-recommendation-events':
    displayName: "Kubernetes Recommendations"
    botkube/kubernetes:
      context: &default-plugin-context
        defaultNamespace: "default"
        rbac:
          group:
            type: Static
            prefix: ""
            static:
              values: [*static-group-name]
      enabled: true
      config:
        log:
          level: debug
        namespaces:
          include:
            - ".*"
        recommendations:
          pod:
            noLatestImageTag: true
            labelsSet: true
          ingress:
            backendServiceValid: true
            tlsSecretValid: true
  'k8s-all-events':
    displayName: "Kubernetes Info"
    botkube/kubernetes:
      context: *default-plugin-context
      enabled: true
      config:
        log:
          level: debug
        filters:
          objectAnnotationChecker: true
          nodeEventsChecker: true
        namespaces:
          include:
            - ".*"
        event:
          types:
            - error
          reason:
            include: []
            exclude: []
          message:
            include: []
            exclude: []

        annotations: {}
        labels: {}

        resources:
          - type: v1/pods
            namespaces:
              include:
                - ".*"
              exclude:
                - some_namespace
            event:
              types:
                - error
          - type: v1/services
            namespaces:
              include:
                - ".*"
              exclude:
                - some_namespace
          - type: networking.k8s.io/v1/ingresses
            namespaces:
              include:
                - ".*"
              exclude:
                - some_namespace
          - type: v1/nodes
            namespaces:
              include:
                - ".*"
              exclude:
                - some_namespace
          - type: v1/namespaces
          - type: v1/persistentvolumes
            namespaces:
              include:
                - ".*"
              exclude:
                - some_namespace
          - type: v1/persistentvolumeclaims
            namespaces:
              include:
                - ".*"
              exclude:
                - some_namespace
          - type: v1/configmaps
            event:
              types:
                - create
                - error
                - delete
            namespaces:
              include:
                - ".*"
              exclude:
                - some_namespace
          - type: rbac.authorization.k8s.io/v1/roles
            namespaces:
              include:
                - ".*"
              exclude:
                - some_namespace
          - type: rbac.authorization.k8s.io/v1/rolebindings
            namespaces:
              include:
                - ".*"
              exclude:
                - some_namespace
          - type: rbac.authorization.k8s.io/v1/clusterrolebindings
            namespaces:
              include:
                - ".*"
              exclude:
                - some_namespace
          - type: rbac.authorization.k8s.io/v1/clusterroles
            namespaces:
              include:
                - ".*"
              exclude:
                - some_namespace
          - type: apps/v1/daemonsets
            namespaces:
              include:
                - ".*"
              exclude:
                - some_namespace
            event:
              types:
                - create
                - update
                - delete
                - error
            updateSetting:
              includeDiff: true
              fields:
                - spec.template.spec.containers[*].image
                - status.numberReady
          - type: batch/v1/jobs
            namespaces:
              include:
                - ".*"
              exclude:
                - some_namespace
            event:
              types:
                - error
            updateSetting:
              includeDiff: true
              fields:
                - spec.template.spec.containers[*].image
                - status.conditions[*].type
          - type: apps/v1/deployments
            namespaces:
              include:
                - ".*"
              exclude:
                - some_namespace
            event:
              types:
                - error
            updateSetting:
              includeDiff: true
              fields:
                - spec.template.spec.containers[*].image
                - status.availableReplicas
          - type: apps/v1/statefulsets
            namespaces:
              include:
                - ".*"
              exclude:
                - some_namespace
            event:
              types:
                - create
                - update
                - delete
                - error
            updateSetting:
              includeDiff: true
              fields:
                - spec.template.spec.containers[*].image
                - spec.replicas
  'k8s-all-err-events':
    displayName: "Kubernetes Errors"
    botkube/kubernetes:
      context:
        defaultNamespace: "default"
        rbac:
          group:
            type: Static
            static:
              values: [ "botkube-plugins-default" ]
      enabled: true
      config:
        log:
          level: debug
        extraButtons:
          - enabled: true
            trigger:
              type: ["error"]
            button:
              displayName: "Get Help"
              commandTpl: "doctor --resource={{ .Kind | lower }}/{{ .Name }} --namespace={{ .Namespace }} --error={{ .Reason }} --bk-cmd-header='AI assistance'"
        namespaces:
          include:
            - ".*"
        event:
          types:
            - error
        resources:
          - type: v1/pods
          - type: v1/services
          - type: v1/nodes
          - type: v1/namespaces
          - type: v1/persistentvolumes
          - type: v1/persistentvolumeclaims
          - type: v1/configmaps
          - type: rbac.authorization.k8s.io/v1/roles
          - type: rbac.authorization.k8s.io/v1/rolebindings
          - type: rbac.authorization.k8s.io/v1/clusterrolebindings
          - type: rbac.authorization.k8s.io/v1/clusterroles
          - type: apps/v1/deployments
          - type: apps/v1/statefulsets
          - type: apps/v1/daemonsets
          - type: batch/v1/jobs

executors:
  k8s-default-tools:
    botkube/kubectl:
      enabled: true
      context: *default-plugin-context
  ai:
    botkube/doctor:
      enabled: true
      config:
        apiKey: "sk-" # Get it here https://platform.openai.com/account/api-keys

plugins:
  cacheDir: "/tmp/plugins"

  repositories:
    botkube:
      url: http://localhost:8080/plugins-dev-index.yaml

settings:
  kubeconfig: /Users/huseyin/.kube/config
  log:
    level: "debug"
  clusterName: "k3d-botkube"
  upgradeNotifier: false

analytics:
  disable: true

configWatcher:
  enabled: false
```
- Run `kubectl run nginx --image=nginx2` to see error event in slack
- Click "Get Help" button to see result.

## Related issue(s)

[#1070](https://github.com/kubeshop/botkube/issues/1070)

### CODE REVIEW

**Review by mszostok — COMMENTED** [review]
Please fix the reported issuse in this PR and the rest issues in the follow-up PR. Thanks!

Follow-up PR:
- add support for `--bk-cmd-header='AI assistance'` in botkube core
- add support for rendering inline buttons and dropdowns in the same line
  - in general if all buttons in a given section doesn't have description then they are printed in a single line. In this case we should also handle the dropdown and print dropdowns also in the line with buttons.

both things were handled on the hackathon PR but there require some polishing: https://github.com/kubeshop/botkube/pull/1066/files

**Comment by huseyinbabal** [comment]
> 

@mszostok I tried to handle them all in this PR, I already included AI assistance partially, so just wanted to wrap up them all. Let's sync tomorrow to visit it once more.

**Comment by huseyinbabal** [comment]
> Please fix the reported issuse in this PR and the rest issues in the follow-up PR. Thanks!
> 
> Follow-up PR:
> 
> * add support for `--bk-cmd-header='AI assistance'` in botkube core
> * add support for rendering inline buttons and dropdowns in the same line
>   
>   * in general if all buttons in a given section doesn't have description then they are printed in a single line. In this case we should also handle the dropdown and print dropdowns also in the line with buttons.
> 
> both things were handled on the hackathon PR but there require some polishing: https://github.com/kubeshop/botkube/pull/1066/files

@mszostok I tried to handle them all in this PR, I already included AI assistance partially, so just wanted to wrap up them all. Let's sync tomorrow to visit it once more.

File: pkg/execute/params.go [hunk 1]
```diff
@@ -140,3 +149,46 @@ func regexFlag(flag, escapedParamVal string) (*regexp.Regexp, error) {
 		escapedParamVal,
 		escapedParamVal))
 }
+
+func extractHeaderParam(cmd string) (string, string, error) {
```
Comment (mszostok) [inline pkg/execute/params.go:153]: we already have generic `extractParam` please use it instead of this 👍

File: pkg/bot/slack_renderer.go [hunk 2]
```diff
@@ -253,30 +258,32 @@ func (b *SlackRenderer) renderContext(in []api.ContextItem) []slack.Block {
...
 
+	out := selects.Elements.ElementSet
+	out = append(out, btns...)
 	return []slack.Block{
 		// We use actions layout as we have only buttons that we want to display in a single line.
 		// https://api.slack.com/reference/block-kit/blocks#actions
 		slack.NewActionBlock(
-			"",
-			btns...,
+			selects.BlockID,
```
Comment (mszostok) [inline pkg/bot/slack_renderer.go:283]: this might be nil and we get a panic, I tried to use kubectl builder and it doesn't work.

**Review by mszostok — COMMENTED** [review]
currently the kubectl builder doesn't work, the renderer logic needs to be adjusted 👍

**Comment by huseyinbabal** [comment]
@mszostok I have applied suggestions and tested accordingly.

Screenshot 2023-07-06 at 16 59 52
Screenshot 2023-07-06 at 17 00 14

Screenshot 2023-07-06 at 16 59 48

File: pkg/execute/params.go [hunk 3]
```diff
@@ -18,6 +18,8 @@ const (
 	paramFlagParseErrorMsg = `incorrect use of %s flag: could not parse flag in %s
 error: %s
 Use %s="value" or %s value`
+	multipleFilters       = "incorrect use of --filter flag: found more than one filter flag"
+	missingCmdFilterValue = `incorrect use of --filter flag: an argument is missing. use --filter="value" or --filter value`
```
Comment (mszostok) [inline pkg/execute/params.go:22]: please remove :)

**Review by mszostok — APPROVED** [review]
awesome work 🚀

**Pull request merged**